aptos-move/framework/src/natives/aggregator_natives/aggregator_v2.rs

```diff
@@ -0,0 +1,232 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::natives::{
+    aggregator_natives::helpers_v2::{
+        aggregator_snapshot_string_info, aggregator_snapshot_u128_info,
+        aggregator_snapshot_u64_info, string_value_info,
+    },
+    AccountAddress,
+};
+use aptos_gas_schedule::gas_params::natives::aptos_framework::*;
+use aptos_native_interface::{
+    safely_pop_arg, RawSafeNative, SafeNativeBuilder, SafeNativeContext, SafeNativeError,
+    SafeNativeResult,
+};
+use move_core_types::value::{MoveStructLayout, MoveTypeLayout};
+use move_vm_runtime::native_functions::NativeFunction;
+use move_vm_types::{
+    loaded_data::runtime_types::Type,
+    values::{Struct, StructRef, Value},
+};
+use smallvec::{smallvec, SmallVec};
+use std::{collections::VecDeque, ops::Deref};
+
+/// The generic type supplied to aggregator snapshots is not supported
+pub const EUNSUPPORTED_AGGREGATOR_SNAPSHOT_TYPE: u64 = 0x02_0005;
+
+/// The aggregator snapshots feature is not enabled
+pub const EAGGREGATOR_SNAPSHOTS_NOT_ENABLED: u64 = 0x02_0006;
+
+/// Checks if the type argument `type_arg` is a string type
+fn is_string_type(context: &SafeNativeContext, type_arg: &Type) -> SafeNativeResult<bool> {
+    let ty = context.deref().type_to_fully_annotated_layout(type_arg)?;
+    if let MoveTypeLayout::Struct(MoveStructLayout::WithTypes { type_, .. }) = ty {
+        if type_.name.as_str() == "String"
+            && type_.module.as_str() == "string"
+            && type_.address == AccountAddress::ONE
+        {
+            return Ok(true);
+        }
+    }
+    Ok(false)
+}
+
+/***************************************************************************************************
+ * native fun create_snapshot(value: Element): AggregatorSnapshot<Element>;
+ **************************************************************************************************/
+
+fn native_create_snapshot(
+    context: &mut SafeNativeContext,
+    ty_args: Vec<Type>,
+    mut args: VecDeque<Value>,
+) -> SafeNativeResult<SmallVec<[Value; 1]>> {
+    if !context.is_aggregator_snapshots_enabled() {
+        return Err(SafeNativeError::Abort {
+            abort_code: EAGGREGATOR_SNAPSHOTS_NOT_ENABLED,
+        });
+    }
+
+    debug_assert_eq!(ty_args.len(), 1);
+    debug_assert_eq!(args.len(), 1);
+    context.charge(AGGREGATOR_V2_CREATE_SNAPSHOT_BASE)?;
+
+    match ty_args[0] {
+        Type::U128 => {
+            let input = safely_pop_arg!(args, u128);
+            Ok(smallvec![Value::struct_(Struct::pack(vec![Value::u128(
+                input
+            )]))])
+        },
+        Type::U64 => {
+            let input = safely_pop_arg!(args, u64);
+            Ok(smallvec![Value::struct_(Struct::pack(vec![Value::u64(
+                input
+            )]))])
+        },
+        _ => {
```

Comment:
I think you can do conditional in the match arm itself here, might be cleaner: https://doc.rust-lang.org/book/ch18-03-pattern-syntax.html#extra-conditionals-with-match-guards